lumicks/pylake/detail/calibrated_images.py

```diff
@@ -0,0 +1,129 @@
+import numpy as np
+import matplotlib.pyplot as plt
+from skimage.measure import block_reduce
+
+
+class CalibratedKymographChannel:
+    """A kymograph image along with its calibration"""
+
+    def __init__(self, name, data, start, time_step, calibration, downsampling_factor=1):
+        self.name = name
+        self.data = data
+        self.start = start
+        self.time_step = time_step
+        self._calibration = calibration
+        self.downsampling_factor = downsampling_factor
+
+    @classmethod
+    def from_array(cls, image, start=0, time_step=int(1e9), calibration=1, downsampling_factor=1):
+        return cls(
+            "test",
+            image,
+            start=start,
+            time_step=time_step,
+            calibration=calibration,
+            downsampling_factor=downsampling_factor,
+        )
+
+    @classmethod
+    def from_kymo(cls, kymo, channel):
+        ts = kymo.timestamps[0, :]
+        time_steps = np.unique(np.diff(ts))
+        if len(time_steps) > 1:
+            raise RuntimeError("Each line should have the same scan time.")
+
+        return cls(
+            kymo.name, getattr(kymo, f"{channel}_image"), ts[0], time_steps[0], kymo.pixelsize_um[0]
+        )
+
+    def _get_rect(self, rect):
+        """Grab a subset of the image in the coordinates of the image"""
+        ((t0, p0), (t1, p1)) = rect
+        t0_pixels, t1_pixels = self.from_seconds(t0), self.from_seconds(t1)
+        p0_pixels, p1_pixels = self.from_coord(p0), self.from_coord(p1)
+
+        if p0_pixels > self.data.shape[0]:
+            raise IndexError(
+                f"Specified minimum position {p0} beyond the valid coordinate range {self.to_coord(self.data.shape[0])}"
+            )
+
+        if t0_pixels > self.data.shape[1]:
+            raise IndexError(
+                f"Specified minimum time {t0} beyond the time range {self.to_seconds(self.data.shape[1])}"
+            )
```

Comment:
We could, but I'm uncertain about this one. There's a performance implication there, since it'd add quite some extra checking when essentially, it's only needed on API boundaries.